<a href="https://colab.research.google.com/github/KondapiPraveen/Machine-Learning-Project/blob/master/MLProject.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

**1.1 Importing Libraries**
 1. Numpy for numerical computations
 2. Pandas for DataFrame Manipulation
 3. Matplotlib for Visulalizations 

In [ ]:
# Loading csv data 
data = pd.read_csv('/content/sample_data/data.csv')

In [ ]:
# Data Overview by printing head of data
data.head()

Make       Model  Year  ... city mpg  Popularity   MSRP
0  BMW  1 Series M  2011  ...       19        3916  46135
1  BMW    1 Series  2011  ...       19        3916  40650
2  BMW    1 Series  2011  ...       20        3916  36350
3  BMW    1 Series  2011  ...       18        3916  29450
4  BMW    1 Series  2011  ...       18        3916  34500

[5 rows x 16 columns]

**1.2 Checking DataTypes and Shape**

Get to know the size of DataFrame and datatypes of data and count values of data in each column

In [ ]:
print(data.shape)
data.info()

(11914, 16)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11914 entries, 0 to 11913
Data columns (total 16 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Make               11914 non-null  object 
 1   Model              11914 non-null  object 
 2   Year               11914 non-null  int64  
 3   Engine Fuel Type   11911 non-null  object 
 4   Engine HP          11845 non-null  float64
 5   Engine Cylinders   11884 non-null  float64
 6   Transmission Type  11914 non-null  object 
 7   Driven_Wheels      11914 non-null  object 
 8   Number of Doors    11908 non-null  float64
 9   Market Category    8172 non-null   object 
 10  Vehicle Size       11914 non-null  object 
 11  Vehicle Style      11914 non-null  object 
 12  highway MPG        11914 non-null  int64  
 13  city mpg           11914 non-null  int64  
 14  Popularity         11914 non-null  int64  
 15  MSRP               11914 non-null  int64  
dtypes: float64

Data Size : Rows = 11914, Columns = 16 <br>
There are some missing values in Engine Fuel Type, Engine HP, Engine Cylinders, Number of Doors, Market Category. <br> 
Market Category, Year are of less priority for estimating the price of car. And other features will be excluded later on analysis.<br>
There are so many classification for single entry in Market Category. It is not feasible to use Market Catefoty for estimation.<br>
Price of a car mostly depends its features not on the year.



**1.3 Null Values**
<br>
Summary Statistics of the data

In [ ]:
data.describe()

Year    Engine HP  ...    Popularity          MSRP
count  11914.000000  11845.00000  ...  11914.000000  1.191400e+04
mean    2010.384338    249.38607  ...   1554.911197  4.059474e+04
std        7.579740    109.19187  ...   1441.855347  6.010910e+04
min     1990.000000     55.00000  ...      2.000000  2.000000e+03
25%     2007.000000    170.00000  ...    549.000000  2.100000e+04
50%     2015.000000    227.00000  ...   1385.000000  2.999500e+04
75%     2016.000000    300.00000  ...   2009.000000  4.223125e+04
max     2017.000000   1001.00000  ...   5657.000000  2.065902e+06

[8 rows x 8 columns]

**STEP 3 : Renaming Columns**
  1. Engine HP => HP
  2. Engine Cylinders => Cylinders
  3. Transmission Type => Transmission
  4. Driven_Wheels => Drive Mode
  5. highway MPG => MPG-H
  6. city MPG => MPG-C
  7. MSRP => Price <br>
  STEP 2 is done later

In [ ]:
data.rename(columns={"Engine HP":"HP","Engine Cylinders":"Cylinders","Transmission Type":"Transmission","Driven_Wheels":"Drive Mode","highway MPG":"MPG-H","city mpg":"MPG-C","MSRP":"Price"},inplace = True)
data

Make       Model  Year  ... MPG-C  Popularity  Price
0          BMW  1 Series M  2011  ...    19        3916  46135
1          BMW    1 Series  2011  ...    19        3916  40650
2          BMW    1 Series  2011  ...    20        3916  36350
3          BMW    1 Series  2011  ...    18        3916  29450
4          BMW    1 Series  2011  ...    18        3916  34500
...        ...         ...   ...  ...   ...         ...    ...
11909    Acura         ZDX  2012  ...    16         204  46120
11910    Acura         ZDX  2012  ...    16         204  56670
11911    Acura         ZDX  2012  ...    16         204  50620
11912    Acura         ZDX  2013  ...    16         204  50920
11913  Lincoln      Zephyr  2006  ...    17          61  28995

[11914 rows x 16 columns]

***STEP 2 : Data Cleaning***
1. Find the appropriate statistical data to replace. <br>
Other columns are not useful estimation but will be kept, In making the model only useful information will be used.<br>
Year is also not useful for estimating price. <br> 
Later correlation matrix and relationship between variables can be seen throught visualizations. 


In [ ]:
data.drop(columns="Market Category")

Make       Model  Year  ... MPG-C  Popularity  Price
0          BMW  1 Series M  2011  ...    19        3916  46135
1          BMW    1 Series  2011  ...    19        3916  40650
2          BMW    1 Series  2011  ...    20        3916  36350
3          BMW    1 Series  2011  ...    18        3916  29450
4          BMW    1 Series  2011  ...    18        3916  34500
...        ...         ...   ...  ...   ...         ...    ...
11909    Acura         ZDX  2012  ...    16         204  46120
11910    Acura         ZDX  2012  ...    16         204  56670
11911    Acura         ZDX  2012  ...    16         204  50620
11912    Acura         ZDX  2013  ...    16         204  50920
11913  Lincoln      Zephyr  2006  ...    17          61  28995

[11914 rows x 15 columns]

**4.1 Duplicate Data Points**
1. It can be seen after removing Marketing Category there are 715 duplicate rows. Keeping first data entry intact
2. Next, we dropped the duplicate entries keeping the first one entry.

In [ ]:
print(data.duplicated(keep='first').value_counts())
data.drop_duplicates(keep='first',inplace=True)

False    11199
True       715
dtype: int64


In [ ]:
Fuel_mode = data["Engine Fuel Type"].mode().values[0]
Hp_med = data["HP"].median()
Cy_mode = data["Cylinders"].mode().values[0]
tr_mode = data["Transmission"].mode().values[0]
nd_mode = data["Number of Doors"].mode().values[0]

**4.2 Imputing Data**
1. Fuel Type is categorical so the most frequent value is chosen for replacement
2. HP is numerical but for robust analysis median is chosen, so outliers don't effect.
3. Cylinders is numerical but by intuation choosing most frequent value is clear.
4. Transmission and Number of doors follow most frequent values is obvious

For some missing numerical values median and mean are close to each other. For some mode is choosen to make more sense 


In [ ]:
data["Engine Fuel Type"].fillna(Fuel_mode,inplace=True)
data["HP"].fillna(Hp_med,inplace=True)
data["Cylinders"].fillna(Cy_mode,inplace=True)
data["Transmission"].fillna(tr_mode,inplace=True)
data["Number of Doors"].fillna(nd_mode,inplace=True)
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 11199 entries, 0 to 11913
Data columns (total 16 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Make              11199 non-null  object 
 1   Model             11199 non-null  object 
 2   Year              11199 non-null  int64  
 3   Engine Fuel Type  11199 non-null  object 
 4   HP                11199 non-null  float64
 5   Cylinders         11199 non-null  float64
 6   Transmission      11199 non-null  object 
 7   Drive Mode        11199 non-null  object 
 8   Number of Doors   11199 non-null  float64
 9   Market Category   7823 non-null   object 
 10  Vehicle Size      11199 non-null  object 
 11  Vehicle Style     11199 non-null  object 
 12  MPG-H             11199 non-null  int64  
 13  MPG-C             11199 non-null  int64  
 14  Popularity        11199 non-null  int64  
 15  Price             11199 non-null  int64  
dtypes: float64(3), int64(5), object(8)
memor

**Plotting Visualizations** <br>
Plots will be plotted between price and all other variables, to see the relationship between them

In [ ]:
# Boxplot for prices
import seaborn as sns
sns.boxplot(x=data["Price"])